File: packages/dapp-browser/android/src/main/java/ee/forgr/capacitor_inappbrowser/WebViewRegistry.java
package ee.forgr.capacitor_inappbrowser;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Singleton registry of all open dApp browser instances on Android, keyed
 * by id. Mirrors the iOS WebViewRegistry.
 *
 * Each instance is a {@link WebViewDialog}. Unlike iOS (which needs the
 * UIWindow Architecture A refactor in chunk 3 to host multiple webviews
 * simultaneously), Android Dialog already supports multiple concurrent
 * instances natively — each Dialog has its own Window and can be shown
 * or hidden independently. The plugin therefore just needs to track
 * which dialog corresponds to which id.
 *
 * The default single-instance code path registers under
 * {@link #DEFAULT_INSTANCE_ID} so legacy callers (faucet-webview,
 * native-notifications, the wallet's PR-3 dApp browser) keep working.
 */
public class WebViewRegistry {
    public static final String DEFAULT_INSTANCE_ID = "default";

    private static final WebViewRegistry INSTANCE = new WebViewRegistry();

    public static WebViewRegistry getShared() {
        return INSTANCE;
    }

    // LinkedHashMap so iteration / allIds() preserves registration order.
    private final Map<String, WebViewDialog> instances = new LinkedHashMap<>();

    private WebViewRegistry() {
    }

    public synchronized void register(String id, WebViewDialog dialog) {
        instances.put(id, dialog);
    }

    public synchronized WebViewDialog get(String id) {
        return instances.get(id);
    }

    public synchronized void remove(String id) {
        instances.remove(id);
    }

    public synchronized List<String> allIds() {
        return new ArrayList<>(instances.keySet());
    }

    public synchronized int count() {
        return instances.size();
    }

    /** Snapshot of every (id, dialog) pair so the caller can iterate without
     *  holding the lock. */
    public synchronized List<Map.Entry<String, WebViewDialog>> snapshot() {
        return new ArrayList<>(instances.entrySet());
    }

    public synchronized void clear() {
        instances.clear();
    }
}

File: packages/dapp-browser/ios/Sources/InAppBrowserPlugin/WebViewRegistry.swift
import Foundation
import UIKit

/// Singleton registry of all open dApp browser instances, keyed by id.
///
/// PR-4 of the Miden Wallet's wechat-style dApp browser plan grows the
/// plugin from a single-instance model (one `WKWebViewController?` field)
/// to a multi-instance model where the wallet can keep several dApps live
/// in parallel and switch between them instantly.
///
/// The registry is the central index. Each instance is a `WKWebViewInstance`
/// holding a `WKWebViewController` plus the per-instance UIWindow / state
/// added in chunks 3+. The default single-instance code path registers the
/// active webview under the id `"default"` so existing callers
/// (`faucet-webview`, `native-notifications`, the legacy `InAppBrowser` JS
/// API) keep working unchanged.
@objc public class WebViewRegistry: NSObject {
    @objc public static let shared = WebViewRegistry()

    /// The id used by the legacy single-instance code path. JS callers that
    /// don't pass an explicit `id` parameter map to this id internally.
    @objc public static let defaultInstanceId = "default"

    private let lock = NSLock()
    private var instances: [String: WKWebViewInstance] = [:]

    private override init() {
        super.init()
    }

    /// Register an instance under the given id, replacing any existing entry.
    @objc public func register(_ instance: WKWebViewInstance) {
        lock.lock()
        defer { lock.unlock() }
        instances[instance.id] = instance
    }

    /// Look up an instance by id. Returns nil if no instance is registered.
    @objc public func get(id: String) -> WKWebViewInstance? {
        lock.lock()
        defer { lock.unlock() }
        return instances[id]
    }

    /// Remove an instance from the registry. Caller is responsible for
    /// teardown of the underlying WKWebViewController / UIWindow.
    @objc public func remove(id: String) {
        lock.lock()
        defer { lock.unlock() }
        instances.removeValue(forKey: id)
    }

    /// Snapshot of the current ids in registration order. Used by
    /// `listInstances` and the LRU eviction logic in PR-6.
    @objc public func allIds() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        return Array(instances.keys)
    }

    /// Number of currently registered instances.
    @objc public var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return instances.count
    }

    /// Iterate over all registered instances. The closure receives
    /// (id, instance) for each entry. Locking is held for the duration of
    /// the iteration so callers must keep the closure short and non-
    /// blocking — defer any heavyweight work outside the closure.
    @objc public func forEach(_ block: (String, WKWebViewInstance) -> Void) {
        lock.lock()
        defer { lock.unlock() }
        for (id, instance) in instances {
            block(id, instance)
        }
    }

    /// Remove ALL registered instances. Caller is responsible for tearing
    /// down the underlying webviews / windows. Used by `closeAll` and on
    /// app shutdown.
    @objc public func removeAll() {
        lock.lock()
        defer { lock.unlock() }
        instances.removeAll()
    }
}
